### ISSUE TITLE
Support usage of the same CA certificate for multiple tenants

### ISSUE THREAD

**Issue by alinaserg** [eclipse/hono#2981]
Eclipse Hono supports device authentication based on device certificates. For this authentication mechanism a CA certificate (trust-anchor) must be configured on the tenant level. As of today CA certificate must be unique among all tenants provided in the DB, as CA certificate is also used to identify tenant-id when device provides its client certificate.

It works fine till this CA-certificate uniqueness can be guaranteed. There are some use cases when device certificates can be issued by the same CA, but devices belong to different tenants. E.g. some produced devices are used only for test purposes and belong to the test tenant and some devices are used in the field and belong to the production tenant. In this use case the differentiation between production and test tenants with the help of CA certificate is not possible, if all these devices use client certificates issued by the same CA.

For such scenarios I think it will be helpful if we have a possibility to identify tenant-id not only on the CA from the client certificate, but using some additional information about tenant-id from the device. E.g. test and production devices use client certificates issued by the same CA and provide tenant-id to which the device belongs to in SNI. To identify tenant-id the information from SNI is used and in addition to that we validate if the client certificate is issued by the correct CA, preconfigured on tenant level. If there's no tenant information provided in SNI, then CA-certificate must be used for tenant identification.

It will provide additional flexibility and cover additional business use cases.

WDYT @sophokles73 @kaniyan @b-abel @calohmn ?

**Comment by sophokles73** [comment]
We are currently using [SNI on the adapters](https://www.eclipse.org/hono/docs/admin-guide/secure_communication/#server-name-indication-sni) to allow devices to use a tenant specific FQDN to connect to the adapter(s).
For example, if a tenant has registered domain `.my-domain.com` and has created a certificate including DNS name `mqtt.my-domain.com` which is put into the MQTT adapter's key store, then the tenant's devices can use `mqtt.my-domain.com` to connect to the MQTT adapter if the devices include the host name in the TLS SNI extension. In this case SNI is used in the adapter to determine a server certificate that matches the host name used by the device.

In the use case described by @alinaserg SNI could be used to serve an additional purpose, which is to determine the identifier of the tenant that the device belongs to. This would require the registration of wild card certificates for tenant specific domain names. In the example given above, the certificate would need to contain `*.my-domain.com` so that it would match all FQDNs that end in `my-domain.com`. Based on this, a device could then include the tenant identifier in the host name included in the SNI extension, e.g. `my-tenant-id.mqtt.my-domain.com`. 

IMHO the adapter should then be able to use SNI to
1. determine a matching server certificate during TLS negotiation,
2. use the prefix from the host name (`my-tenant-id`) to determine the tenant that the device (presumably) belongs to and
3. look up the trust anchors registered for the tenant and validate the client certificate using the tenant's trust anchor.

If the device does not include a corresponding SNI extension value, then Hono's default mechanism can be used to determine the tenant based on the client certificate's issuer DN.

@calohmn @kaniyan WDYT?

**Comment by kaniyan** [comment]
With this feature, a device should be able to authenticate with a _client certificate_ and _tenant-id_ provided as part of SNI. My understanding is that SNI can be modified anytime to use different tenants whereas those _device certificates might remain the same_. After we get the _tenant-id_ from SNI (as described in the above comment), how we find the correct trust anchor can be challenging. I am not an expert in SNI setup, please correct me if I am wrong. My understanding is that the device certificate's _issuer-dn_ will be further used to look up the correct trust anchor along with the _tenant-id_. As part of this feature, we should allow more than one tenants to have the same trust anchors. In the current Hono setup, there is a unique constraint on trusted anchor's _subject-dn_. It means,  no two tenants can have same trust anchors. FMPOV we cannot just extend that constraint to allow multiple tenants to have same trust anchors. In that case, when no _tenant-id_ is provided or no SNI is configured, we cannot identify the tenant as there will be more than one trust anchors matching the _issuer-dn_ of the device certificate. It might not be simple to extend the unique constraint and may need a different approach.

**Comment by sophokles73** [comment]
> FMPOV we cannot just extend that constraint to allow multiple tenants to have same trust anchors. In that case, when no tenant-id is provided or no SNI is configured, we cannot identify the tenant as there will be more than one trust anchors matching the issuer-dn of the device certificate. 

Good point

**Comment by boaks** [comment]
On additional point, which may require further analysis (especially in the client implementation):
Libraries do frequently just send the DNS hostname in the SNI. Without "advanced access", that would require to setup DNS entries for tenants using that feature.

**Comment by sophokles73** [comment]
@kaniyan thanks for the detailed proposal :-) One of the advantages of your approach would certainly be that we can simply keep the unique constraints as they are. I am a little concerned regarding the implications of introducing a parent-child relation between tenants. We would need to define the semantics of this relationship in detail, covering all aspects of authorization and configuration etc. As you pointed out, it will also make queries more complex as we would then need to take care of depth and potential circular dependencies.

Maybe we can use an approach that is little more lightweight but would require us to make some adjustments regarding the uniqueness constraints. We could, for example introduce a *trust-anchor-group* property to the tenant object which can be used to implicitly define a group of tenants that are allowed to share trust anchors and which therefore require devices to use SNI to indicate the tenant ID that they belong to.

We would then need to replace the unique constraints with explicit checks when creating and updating tenants. However, these checks should be as simple as: 

Is there another tenant that contains the same trust anchor but does not have the same *trust-anchor-group* value?

We use a similar approach for creating implicit groups of (gateway) devices already.
With such an approach we could keep all existing semantics as they are and would not need to think about inheritance between tenants.

**Comment by kaniyan** [comment]
> We could, for example introduce a _trust-anchor-group_ property to the tenant object which can be used to implicitly define a group of tenants that are allowed to share trust anchors and which therefore require devices to use SNI to indicate the tenant ID that they belong to.

Do you mean as in the below example? The tenants having `trust-anchor-group` as _ACME_ share the trust anchors among themselves. Since we already have a similar approach in case of gateway devices and also no need to introduce a new relationship among tenants, I prefer this approach too. But I think we still need to return all the trust anchors of the group during _get Tenant_ operation. For example `Tenant.getTrustedCertificateAuthoritySubjectDNs()` should be able to return all the subject-dns from the group and not only that belongs to the particular tenant.

```
{
    "tenant": "ACME-Dev",
    "trust-anchor-group": "ACME"
}

{
    "tenant": "ACME-Test",
    "trust-anchor-group": "ACME"
}
```

The attributes such as `auto-provisioning-enabled`, `auto-provisioning-device-id-template` are attached to the trust Anchor. Since the trust anchors are being shared and it means that these attributes are also being shared among the tenants belonging to the same group.


> which therefore require devices to use SNI to indicate the tenant ID that they belong to.

What if there is no tenant ID provided in the SNI? In that case more than one trust anchors will be matched. If that is the case and they all belong to a group then we can consider the request invalid and no tenant will be returned and thereby authentication fails.

**Comment by sophokles73** [comment]
> Do you mean as in the below example? 

Yes, but they should, of course, also include the trust anchor specifications.

> The attributes such as auto-provisioning-enabled, auto-provisioning-device-id-template are attached to the trust Anchor. Since the trust anchors are being shared and it means that these attributes are also being shared among the tenants belonging to the same group.

IMHO there is no need to share the values of the auto-provisioning-enabled, auto-provisioning-device-id-template properties among the group. Why can't different tenants define them independently from each other, even if they belong to the same trust-anchor-group? FMPOV the *trust-anchor-group* membership's only purpose is to support enforcing the uniqueness constraint across tenants without requiring a DB constraint.

> What if there is no tenant ID provided in the SNI? In that case more than one trust anchors will be matched. If that is the case and they all belong to a group then we can consider the request invalid and no tenant will be returned and thereby authentication fails.

Yes, in that case authentication fails.

**Comment by kaniyan** [comment]
> IMHO there is no need to share the values of the auto-provisioning-enabled, auto-provisioning-device-id-template properties among the group. Why can't different tenants define them independently from each other, even if they belong to the same trust-anchor-group? FMPOV the _trust-anchor-group_ membership's only purpose is to support enforcing the uniqueness constraint across tenants without requiring a DB constraint.

Now to me it sounds like that the _trust-anchor-group_ attribute is mainly there to deal with the uniqueness constraint and actually the trust anchors are not being shared. Let us take the below example. What will happen when a device tries to connect with a certificate with _issuer-dn_  "CN=ACME,O=Eclipse,OU=Hono" and tenant-id in SNI as _ACME-Test_? Will the trust anchor from _ACME-Dev_ be used as the both tenants belong to the same group? Or the _ACME-Test_ should also contain a trusted anchor with with _issuer-dn_  "CN=ACME,O=Eclipse,OU=Hono"?

```
 {
    "tenant": "ACME-Dev",
    "trust-anchor-group": "ACME",
    "trusted-ca": [
        {
            "subject-dn": "CN=ACME,O=Eclipse,OU=Hono",
            "public-key": "MIICIjA...==",
            "not-before": "2020-09-18T08:35:40Z",
            "not-after": "2022-09-17T08:35:40Z"
        }
    ]
}

{
    "tenant": "ACME-Test",
    "trust-anchor-group": "ACME"
}
```

**Comment by sophokles73** [comment]
> Will the trust anchor from ACME-Dev be used as the both tenants belong to the same group? Or the ACME-Test should also have a trusted anchor with issuer-dn "CN=ACME,O=Eclipse,OU=Hono"?

The latter. Based on your example configuration, the device should not be possible to authenticate successfully because the client's certificate cannot be validated because the ACME-Test tenant does not define any trust anchors.

**Comment by sophokles73** [comment]
closed via #3054
